src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleNull.cs

```diff
@@ -8,44 +8,55 @@ namespace System.Text.Json
 {
     public static partial class JsonSerializer
     {
-        private static bool HandleNull(ref Utf8JsonReader reader, ref ReadStack state, JsonSerializerOptions options)
+        private static bool HandleNull(ref Utf8JsonReader reader, ref ReadStack state)
         {
             if (state.Current.SkipProperty)
             {
                 return false;
             }
 
-            // If null is read at the top level and the type is nullable, then the root is "null" so just return.
-            if (reader.CurrentDepth == 0 && (state.Current.JsonPropertyInfo == null || state.Current.JsonPropertyInfo.CanBeNull))
+            JsonPropertyInfo jsonPropertyInfo = state.Current.JsonPropertyInfo;
+
+            if (jsonPropertyInfo == null || (reader.CurrentDepth == 0 && jsonPropertyInfo.CanBeNull))
             {
                 Debug.Assert(state.IsLastFrame);
                 Debug.Assert(state.Current.ReturnValue == null);
                 return true;
             }
 
-            JsonPropertyInfo propertyInfo = state.Current.JsonPropertyInfo;
-            if (!propertyInfo.CanBeNull)
-            {
-                ThrowHelper.ThrowJsonException_DeserializeCannotBeNull(reader, state.JsonPath);
-            }
+            Debug.Assert(jsonPropertyInfo != null);
 
-            if (state.Current.IsEnumerable || state.Current.IsDictionary || state.Current.IsIDictionaryConstructible)
+            if (state.Current.IsCollectionForClass)
             {
-                ApplyObjectToEnumerable(null, ref state, ref reader);
+                AddNullToCollection(jsonPropertyInfo, ref reader, ref state);
                 return false;
             }
 
-            if (state.Current.IsEnumerableProperty || state.Current.IsDictionaryProperty || state.Current.IsIDictionaryConstructibleProperty)
+            if (state.Current.IsCollectionForProperty)
             {
-                bool setPropertyToNull = !state.Current.CollectionPropertyInitialized;
-                ApplyObjectToEnumerable(null, ref state, ref reader, setPropertyDirectly: setPropertyToNull);
+                bool addElement = state.Current.CollectionPropertyInitialized;
 
-                if (setPropertyToNull)
+                if (addElement)
```

Comment:
why store this if used only once?

Reply:
It was for readability -- `state.Current.CollectionPropertyInitialized;` is not too descriptive. I will add a comment instead. Thanks